Sayılar (bilgiler) bilgisayarda ikili sayılar (binary) olarak temsil edilir.

Taban aritmetiği ile ilgili ayrıntılı bilgi:   
http://www.cplusplus.com/doc/hex/   
http://www.kpsskonu.com/genel-yetenek/matematik/taban-aritmetigi/   


In [1]:
#include <iostream>
#include <bitset>
using namespace std;

In [4]:
int a;

In [9]:
a=12;

a sayısı bellekte 2 tabanında sayı olarak saklanır:

$(12)_{10}= (1100)_2$

10 tabanında 572 sayısının değeri:  
$  (5 \times 10 ^3) + (7 \times 10^1) + (2 \times 10^0) $ olarak hesaplanır.

Benzer şekilde  2 tabanındaki 10011 sayısının değeri:   
$  (1 \times 2 ^4) +(0 \times 2 ^3) +(0 \times 2 ^2) +(1 \times 2 ^1) +(1 \times 2 ^0)  $ olarak hesaplanır.   
Bu durumda: $(10011)_2=(19)_{10}$

Bilgisayarda çok sık kullanılan  8 sayı tabanı (octal), 16 sayı tabanı (hexadecimal) sayı tabanları vardır.   

Octal sayı tabanında rakamlar 0-7 arasındadır:  
$(2743)_8=(2 \times 8 ^3) +(7 \times 8 ^2) +(4 \times 8 ^1) +(3 \times 8 ^0) =(1507)_{10}$

Hexadecimal sayı tabanında rakamlar 0-15 arasındadır. 10,11,12,13,14, ve 15 rakamları için sırasıyla A,B,C,D,E,F harfleri kullanılır:
$(A3D8)_{16}=(A \times 16 ^3) +(3 \times 16 ^2) +(D \times 16 ^1) +(8 \times 16 ^0) = \\ ~~~~~~~~~~~~~~~~~~~~(10 \times 16 ^3) +(3 \times 16 ^2) +(13 \times 16 ^1) +(8 \times 16 ^0)=(41944)_{10}$

In [21]:
//on tabanındaki sayıların hex ve oct gösterimleri
cout << hex << 41944 << endl;
cout << oct << 1507 << endl;

//hex ve oct tabanındaki sayıların on tabanında gösterimleri
cout << dec << 0xa3d8 << endl;
cout << dec << 02743 << endl;


a3d8
2743
41944
1507


In [25]:
for(a=0; a<30; a++)
 cout << dec << a <<":"<<hex << a << " ";  
    

0:0 1:1 2:2 3:3 4:4 5:5 6:6 7:7 8:8 9:9 10:a 11:b 12:c 13:d 14:e 15:f 16:10 17:11 18:12 19:13 20:14 21:15 22:16 23:17 24:18 25:19 26:1a 27:1b 28:1c 29:1d 

In [27]:
for(a=0; a<30; a++)
    cout << dec << a <<":"<<oct << a << " ";

0:0 1:1 2:2 3:3 4:4 5:5 6:6 7:7 8:10 9:11 10:12 11:13 12:14 13:15 14:16 15:17 16:20 17:21 18:22 19:23 20:24 21:25 22:26 23:27 24:30 25:31 26:32 27:33 28:34 29:35 

**Bitwise İşlemler**   
&: ve   
| : veya  
~ : değil   
^ : xor  
<< : sola kaydırma  
\>> : sağa kaydırma



In [2]:
unsigned int a=12;
unsigned int b=10;

In [3]:
cout << dec << "Dec: "<< a << " Hex: " << uppercase <<hex << a << " Bin: "<<bitset<8>(a) << endl;
cout << dec << "Dec: "<< b << " Hex: " << uppercase <<hex << b << " Bin: "<<bitset<8>(b) << endl;

Dec: 12 Hex: C Bin: 00001100
Dec: 10 Hex: A Bin: 00001010


VE işleminde her iki bit de 1 ise sonuç 1 olur diğer durumlarda sonuç 0 olur.

| a | b | f  |
|---|---|----|
| 0 | 0 | 0  |
| 0 | 1 | 0  |
| 1 | 0 | 0  |
| 1 | 1 | 1  |

In [12]:
cout <<bitset<8>(a) << endl;
cout <<bitset<8>(b) << endl;
cout <<bitset<8>(a&b) << endl;

00001100
00001010
00001000


VEYA işleminde bitlerden 1 tanesinin  1 olması sonucu 1 yapmak için yeterlidir.

| a | b | f  |
|---|---|----|
| 0 | 0 | 0  |
| 0 | 1 | 1  |
| 1 | 0 | 1  |
| 1 | 1 | 1  |

In [14]:
cout <<bitset<8>(a) << endl;
cout <<bitset<8>(b) << endl;
cout <<bitset<8>(a|b) << endl;

00001100
00001010
00001110


XOR işlemi (mod 2)'de toplama işlemi gibi düşünülebilir.

| a | b | f  |
|---|---|----|
| 0 | 0 | 0  |
| 0 | 1 | 1  |
| 1 | 0 | 1  |
| 1 | 1 | 0  |

In [15]:
cout <<bitset<8>(a) << endl;
cout <<bitset<8>(b) << endl;
cout <<bitset<8>(a^b) << endl;

00001100
00001010
00000110


DEĞİL işleminde bit 0 ise 1 1 ise 0 yapılır.

| a | f  |
|---|----|
| 0 | 1  |
| 1 | 0  |

In [16]:
cout <<bitset<8>(a) << endl;
cout <<bitset<8>(~a) << endl;

00001100
11110011


Sağa kaydırma operatörü: >> 

In [18]:
cout <<bitset<8>(a) << endl;
cout <<bitset<8>(a>>1) << endl;
cout <<bitset<8>(a>>2) << endl;

00001100
00000110
00000011


Sola kaydırma operatörü: <<

In [22]:
cout <<bitset<8>(a) << endl;
cout <<bitset<8>(a<<1) << endl;
cout <<bitset<8>(a<<2) << endl;

00001100
00011000
00110000


**Set Bit** işlemi, bir değişkeninin en sağdaki biti 0. biti olmak üzere n. bitini 1 yapma işlemidir.

In [25]:
cout <<bitset<16>(a) << endl;

0000000000001100


a değişkeninin 10. bitini set edelim.

In [5]:
int n;

In [5]:
n=10;
cout << "a: "<<bitset<16>(a) << endl;
cout << "   "<<bitset<16>(a | (1<<n)) << endl;

a: 0000000000001100
   0000010000001100


**Clear Bit** işlemi, bir değişkeninin en sağdaki biti 0. biti olmak üzere n. bitini 0 yapma işlemidir.

a değişkeninin 3. bitini 0 yapalım.


In [6]:
n=3;
cout << "a: "<<bitset<16>(a) << endl;
cout << "   "<<bitset<16>(a & ~(1<<n)) << endl;

a: 0000000000001100
   0000000000000100


**Test Bit** işlemi, bir değişkeninin en sağdaki biti 0. biti olmak üzere n. bitinin 1 veya  0 olduğunu kontrol etme işlemidir.

In [7]:
// a değişkeninin 3. bitini kontrol etme
n=3;
cout << "a: "<<bitset<16>(a) << endl;
cout << "   "<<((a >> n) & 1) << endl;

a: 0000000000001100
   1


In [8]:
// a değişkeninin 4. bitini kontrol etme
n=4;
cout << "a: "<<bitset<16>(a) << endl;
cout << "   "<<((a >> n) & 1) << endl;

a: 0000000000001100
   0


**Toggle Bit** işlemi, bir değişkeninin en sağdaki biti 0. biti olmak üzere n. biti 1 ise 0 yapma 0 ise 1 yapma işlemidir.

In [9]:
n=2;
cout << "a: "<<bitset<16>(a) << endl;
cout << "   "<<bitset<16>(a ^ (1<<n)) << endl;
n=4;
cout << "   "<<bitset<16>(a ^ (1<<n)) << endl;

a: 0000000000001100
   0000000000001000
   0000000000011100


**Clear Range** bir bölgenin temizlenmesi (0 yapılması) işlemidir.     
Temizlenecek kadar $11\cdots11$ sayısı elde etmek gerekir.  
Örneğin 7. bitten itibaren 5 bitlik bir bölge 0 yapılacaksa 11111=0x1F sayısı kullanılır.

In [ ]:
unsigned int cr;

In [32]:
cr=0x1F;

In [33]:
a=7917;
n=7;
cout << "a: "<<bitset<16>(a) << endl;
cout << "   "<<bitset<16>(a & ~(cr<<n)) << endl;

a: 0001111011101101
   0001000001101101


**Set Range** bir bölgeye bir verinin yazılması işlemidir.  
Bu işlemi yapabilmek için önce o bölgenin temizlenmesi gerekmektedir.  
Temzilenme işleminden sonra yazılmak istenen veri VEYA işlemi ile yapılabilir.

Bir değişkenin 9. bitinden itibaren 1010 verisini yazalım:

In [35]:
int d;

In [41]:
a=7917;
cr=0xF;
d=0xA;
n=9;
cout << "a: "<<bitset<16>(a) << endl;
cout << "   "<<bitset<16>((a & ~(cr<<n)) | (d << n)) << endl;

a: 0001111011101101
   0001010011101101
